irods/test/modules/close_data_objects.py

```diff
@@ -0,0 +1,47 @@
+# This helper module can "double" as a Python script, allowing us to run the below
+# test() method either within the current process or in a child process.  The
+# method in question can thus be called by the following unit tests so that we may assert
+# proper data object auto-closing functionality under these respective scenarios:
+#
+#    irods.test.data_obj_test.TestDataObjOps.test_data_objects_auto_close_on_function_exit__issue_456
+#    irods.test.data_obj_test.TestDataObjOps.test_data_objects_auto_close_on_process_exit__issue_456
+
+from __future__ import print_function
+import sys
+import contextlib
+try:
+    import irods.configuration as config
+except:
+    pass
+from datetime import datetime
+import os
+from irods.test import helpers
+
+@contextlib.contextmanager
+def auto_close_data_objects(value):
+    if 'config' not in globals():
+        yield
+        return
+    try:
+        ORIGINAL_VALUE = config.data_objects.auto_close
+        config.data_objects.auto_close = value
+        yield
+    finally:
+        config.data_objects.auto_close = ORIGINAL_VALUE
+
+def test(return_locals = True):
+    with auto_close_data_objects(True):
+        expected_content = 'content'
+        ses = helpers.make_session()
+        name = '/{0.zone}/home/{0.username}/{1}-object.dat'.format(ses, helpers.unique_name(os.getpid(), datetime.now()))
+        f = ses.data_objects.open(name,'w')
+        f.write(expected_content.encode('utf8'))
+        L=locals()
+        # By default, ses and f will be automatically exported to calling frame (with L being returned),
+        # but by specifying a list/tuple of keys we can export only those specific locals by name.
+        return L if not isinstance(return_locals,(tuple,list)) \
+            else [ L[k] for k in return_locals ]
```

Comment:
For simplicity, could we just return `name` and `expected_content` as a tuple? Maybe I'm not seeing the use case for returning *everything*.

Reply:
The use case for returning everything is to perpetuate the call frame's variables' lifetimes, as if the function's code were operating in the top scope. Just a convenience to not need to repeat the test( ) func's code - 6 lines at its core - verbatim.

I have no problem with either
  * a much more thorough-going explanation in the comment
  * duplicating the test( ) function's code in its entirety at the top scope

if it makes the meaning clearer....

Reply:
I just find the amorphous type of the input(s) and output(s) to be a little confusing from a usage perspective, and only really noticed since it's in a separate file which could be used by other tests (not that they'd need to).

Maybe the name of the file and/or class could be explicitly tied to these specific tests so that it's clear that this is not general-purpose? More explanation may be good enough as well, as you've suggested.